## Peer-graded Assignment: Segmenting and Clustering Neighborhoods in Toronto Part One

### 1. Start by creating a new Notebook for this assignment.

In [41]:
#Library import
from bs4 import BeautifulSoup
import requests

### 2. Scrap Wikipedia Page

Use the Notebook to build the code to scrape the following Wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M, in order to obtain the data that is in the table of postal codes and to transform the data into a pandas dataframe like the one shown below:

In [42]:
#Accessing Url
url=requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text

#BeautifulSoup
soup=BeautifulSoup(url,'lxml')
neighborhood_table=soup.find('table',{'class':'wikitable sortable'})

In [43]:
# Search for tr tag
table_rows = neighborhood_table.find_all('tr') 

#Initialize Empty list for storing row data
neighborhood_data=[] 

#Iterate through each row in table_rows and append the it to list
for tr in table_rows:
    td = tr.find_all('td')
    row = [i.text for i in td]
    neighborhood_data.append(row)   

### 3. To create the above dataframe:

- The dataframe will consist of three columns: PostalCode, Borough, and Neighborhood
- Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.
- More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you      will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will      be combined into one row with the neighborhoods separated with a comma as shown in row 11 in the above table.
   If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough.
   Clean your Notebook and add Markdown cells to explain your work and any assumptions you are making.
- In the last cell of your notebook, use the .shape method to print the number of rows of your dataframe.

- Data Frame Creation

In [44]:
#Remove null and empy header
neighborhood_data.remove([ ]) 

In [45]:
#import pandas library
import pandas as pd

#Initialize Empty DataFrame with columne name 'PostalCode','Borough','Neighborhood'
df=pd.DataFrame(columns=['PostalCode','Borough','Neighborhood']) 

#Iterate through each item in list neighborhood_data and append it to data frame 
for each in neighborhood_data:
    df=df.append({'PostalCode': each[0],
                  'Borough':each[1],
                 'Neighborhood':each[2]},ignore_index=True)

Each row in data frame have new line character \n below steps is used to remove it

In [46]:
#Each row values have \n character, replace function use for removing it
#Remove new line character \n from PostalCode, Borough and Neighborhood

df.PostalCode.replace({r'\n$':''},regex=True,inplace=True)

df.Borough.replace({r'\n$':''},regex=True,inplace=True)

df.Neighborhood.replace({r'\n$':''},regex=True,inplace=True)

- Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.

In [47]:
#Deleting all row whose Borough is "Not assigned"
import numpy as np
index_to_drop=np.where(df['Borough']=='Not assigned')
df.drop(df.index[index_to_drop],inplace=True)
df.reset_index(drop=True,inplace=True)

- Display Data Frame

In [48]:
df.head(12)

PostalCode           Borough  \
0         M3A        North York   
1         M4A        North York   
2         M5A  Downtown Toronto   
3         M6A        North York   
4         M7A  Downtown Toronto   
5         M9A         Etobicoke   
6         M1B       Scarborough   
7         M3B        North York   
8         M4B         East York   
9         M5B  Downtown Toronto   
10        M6B        North York   
11        M9B         Etobicoke   

                                         Neighborhood  
0                                           Parkwoods  
1                                    Victoria Village  
2                           Regent Park, Harbourfront  
3                    Lawrence Manor, Lawrence Heights  
4         Queen's Park, Ontario Provincial Government  
5             Islington Avenue, Humber Valley Village  
6                                      Malvern, Rouge  
7                                           Don Mills  
8                     Parkview Hill, Woodbine Gardens  
9                            Garden District, Ryerson  
10                                          Glencairn  
11  West Deane Park, Princess Gardens, Martin Grov...

In [49]:
df.to_csv('Dataset_For_Neighborhoods_in_Toronto.csv')

- Check and Merge Different Neighborhood for Same PostalCode

In [50]:
#Get the unique PostalCode size
df.PostalCode.unique().size

103

In [51]:
#Display Total PostalCode Size
df.PostalCode.size

103

The count of Unique PostalCode is same as that of Total PostalCode (which is 103), Hence there is no duplicate PostalCode is present and need merging of Neighborhood.

In [52]:
np.where(df.PostalCode.duplicated())

(array([], dtype=int64),)

- Shape

In [53]:
df.shape

(103, 3)

##### End of Part one!!!